Spring Boot Cloud Config

lony2003 opened the pull request:
# Description

A new library that implement a backend of Spring Cloud Config.

Originally from https://github.com/fangkehou-team/dapr-spring, this library created a backend of SpringCloudConfig just like SpringCloudVault.
The original library only uses secret store as config store api is not stable at that time.
As the configuration api is stable now, the config loader using that api has been implemented, but subscription of configuration has not implemented, just for reservation.

## Issue reference

this PR will close: #1225 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Extended the documentation

artursouza committed:
Update springboot to latest minor.patch version. (#826)

(cherry picked from commit 31a47f55212ad0331562d6b308b938d111fe60ae)
Signed-off-by: lony2003 <zhangke200377@outlook.com>

lony2003 committed:
style(springboot autoconfigure): use a PROPERTY_PREFIX instead of hardcoded prefix(dapr#1225)

create a public static final value PROPERTY_PREFIX in DaprClientProperties
change the prefix value in the ConfigurationProperties to DaprClientProperties.PROPERTY_PREFIX
Dapr Cloud Config rely on that.

Signed-off-by: lony2003 <zhangke200377@outlook.com>

lony2003 committed:
feat(springboot cloudconfig): A new library that implement a backend of Spring Cloud Config(dapr#1225)

Originally from https://github.com/fangkehou-team/dapr-spring, this library created a backend of SpringCloudConfig just like SpringCloudVault.
The original library only uses secret store as config store api is not stable at that time.
As the configuration api is stable now, the config loader using that api would be implemented later.

Signed-off-by: lony2003 <zhangke200377@outlook.com>

salaboy commented:
@lony2003 first of all, thanks so so much for contributing back with this. This is highly appreciated.. I will start adding comments in the PR with small details that we need to fix and questions about how this would work. 

It will be great if you can add as part of this PR an example on the `spring-boot-examples/` directory showing how this functionality would work. We probably also need tests inside `sdk-tests` to demonstrate that this works from a client point of view. You will notice that we are heavily relying on Testcontainers for all of this, so it will be quite cool to see if the testcontainers integration is missing something to support this functionality.

salaboy commented on dapr-spring/dapr-spring-cloudconfig/src/main/java/io/dapr/spring/boot/cloudconfig/autoconfigure/DaprCloudConfigAutoConfiguration.java:
```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright (c) 2016-2024 Team Fangkehou
```
@lony2003 you need to use the standard Dapr header for all the new files. https://github.com/dapr/java-sdk/blob/master/dapr-spring/dapr-spring-boot-autoconfigure/src/main/java/io/dapr/spring/boot/autoconfigure/client/DaprClientAutoConfiguration.java#L1

salaboy commented on dapr-spring/dapr-spring-cloudconfig/src/main/java/io/dapr/spring/boot/cloudconfig/config/CloudConfigPropertiesDaprConnectionDetails.java:
```diff
@@ -0,0 +1,46 @@
+import io.dapr.spring.boot.autoconfigure.client.DaprClientProperties;
+import io.dapr.spring.boot.autoconfigure.client.DaprConnectionDetails;
+
+class CloudConfigPropertiesDaprConnectionDetails implements DaprConnectionDetails {
```
@lony2003 do we need to duplicate the DaprConnectionDetails, why are these different for the normal DaprConnectionDetails?

lony2003 replied:
No difference, just because the PropertyDaprConnectionDetails is protected and can't be used in another package.

salaboy commented on dapr-spring/dapr-spring-cloudconfig/src/main/java/io/dapr/spring/boot/cloudconfig/config/DaprCloudConfigClientManager.java:
```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (c) 2016-2024 Team Fangkehou
```
@lony2003 same as before

salaboy commented on dapr-spring/dapr-spring-cloudconfig/src/main/java/io/dapr/spring/boot/cloudconfig/config/DaprCloudConfigProperties.java:
```diff
@@ -0,0 +1,54 @@
+ * The properties for creating dapr client.
+ */
+@ConfigurationProperties(DaprCloudConfigProperties.PROPERTY_PREFIX)
+public class DaprCloudConfigProperties {
```
@lony2003 I think having separate properties makes sense as this is a spring cloud feature.. so it is ok to align with that.

lony2003 commented:
> @lony2003 first of all, thanks so so much for contributing back with this. This is highly appreciated.. I will start adding comments in the PR with small details that we need to fix and questions about how this would work.
> 
> It will be great if you can add as part of this PR an example on the `spring-boot-examples/` directory showing how this functionality would work. We probably also need tests inside `sdk-tests` to demonstrate that this works from a client point of view. You will notice that we are heavily relying on Testcontainers for all of this, so it will be quite cool to see if the testcontainers integration is missing something to support this functionality.

Thanks. the code is not completed currently, I created this pull request is for the problems in issue dapr#1225 , I think we need a schema for cloud config url, and have some other things to be discussed😊😊😊

lony2003 committed:
style(springboot cloudconfig): update file header

Signed-off-by: lony2003 <zhangke200377@outlook.com>

lony2003 committed:
style(springboot cloudconfig): change code style to match checkstyle

Signed-off-by: lony2003 <zhangke200377@outlook.com>

lony2003 commented:
# Schemas

## Secret Store Component

### multiValued = false:

#### store content(file secret store as example)

```json
{
	"dapr.spring.demo-config-secret.singlevalue": "testvalue",
	"multivalue-properties": "dapr.spring.demo-config-secret.multivalue.v1=spring\ndapr.spring.demo-config-secret.multivalue.v2=dapr",
	"multivalue-yaml": "dapr:\n  spring:\n    demo-config-secret:\n      multivalue:\n        v3=cloud"
}
```

#### valid demo url

- `dapr:secret:democonfig/multivalue-properties?type=doc`
- `dapr:secret:democonfig/multivalue-yaml?type=doc`
- `dapr:secret:democonfig/dapr.spring.demo-config.singlevalue?type=value`
- `dapr:secret:democonfig?type=single-value`
- `dapr:secret:democonfig?type=single-doc`

### multiValued = true, nestedSeparator = ".":

#### store content(file secret store as example)

```json
{
	"value1": {
		"dapr": {
			"spring": {
				"demo-config-secret": {
					"multivalue": {
						"v4": "config"
					}
				}
			}
		}
	}
}
```

#### valid demo url
- `dapr:secret:demo-config-multi/value1?type=multi`
- `dapr:secret:demo-config-multi?type=multi`

|  parameter  | description | default | available |
|--------------------|--------------------|--------------------|--------------------|
| type | value type | `value`/`single-value` | `value`/`doc`/`single-value`/`single-doc`/`multi` |

## Configuration Component

#### store content(table as example)

| key | value |
|--------------------|--------------------|
| dapr.spring.demo-config-config.singlevalue | testvalue |
| multivalue-properties | `dapr.spring.demo-config-config.multivalue.v1=spring\ndapr.spring.demo-config-config.multivalue.v2=dapr` |
| multivalue-yaml | `dapr:\n  spring:\n    demo-config-config:\n      multivalue:\n        v3=cloud` |

#### valid demo url

- `dapr:config:democonfigconf/dapr.spring.demo-config-config.singlevalue?type=value`
- `dapr:config:democonfigconf/multivalue-properties?type=doc`
- `dapr:config:democonfigconf/multivalue-yaml?type=doc`
- `dapr:config:democonfigconf?type=doc`
- `dapr:config:democonfigconf?type=value`

|  parameter  | description | default | available |
|--------------------|--------------------|--------------------|--------------------|
| type | value type | `value` | `doc`/`value` |
| subscribe | subscribe this configuration | `false` | `true`/`false` |

lony2003 committed:
feat(springboot cloudconfig): Implement Configuration Cloud Config Importer

Implemented the configuration importer using dapr client configuration api
also done a style check

Signed-off-by: lony2003 <zhangke200377@outlook.com>

lony2003 committed:
fix(springboot cloudconfig): Update version of dapr-spring-parent

Signed-off-by: lony2003 <zhangke200377@outlook.com>

lony2003 committed:
feat(springboot cloudconfig): test cases for cloud config

Signed-off-by: lony2003 <zhangke200377@outlook.com>

lony2003 committed:
fix(springboot cloudconfig): Fix problems of PropertySourceLoader

Springboot‘s PropertySourceLoader will return a invalid source if it doesn't support that instead of throw errors.
So we have to clear that what kind of loader we want to use, that is , that kind of resource it is.

Signed-off-by: lony2003 <zhangke200377@outlook.com>

lony2003 committed:
feat(springboot cloudconfig): Change the behavior of Config Type

Enum of Config Type has lack of file extension suport, reimplement it as class

Signed-off-by: lony2003 <zhangke200377@outlook.com>

lony2003 committed:
test(springboot cloudconfig): Implemented integration test for secret and config store

Signed-off-by: lony2003 <zhangke200377@outlook.com>

lony2003 committed:
test(springboot cloudconfig): fix integration test for secret and config store (try1)

Signed-off-by: lony2003 <zhangke200377@outlook.com>

lony2003 committed:
docs(springboot cloudconfig): Write examples and docs for springboot cloudconfig

Signed-off-by: lony2003 <zhangke200377@outlook.com>

lony2003 commented:
All jobs have done, waiting for review. Note that subscription of configuration api is not implemented, just as a reservation.

salaboy commented on dapr-spring/dapr-spring-cloudconfig/pom.xml:
```diff
@@ -0,0 +1,91 @@
+            <scope>compile</scope>
+        </dependency>
+
+<!--        <dependency>-->
```
@lony2003 we can remove the comments.. there is no need to keep them

lony2003 replied:
removed.

salaboy commented on dapr-spring/dapr-spring-cloudconfig/src/test/java/io/dapr/spring/boot/cloudconfig/CloudConfigTestApplication.java:
```diff
@@ -0,0 +1,14 @@
+import org.springframework.context.annotation.ComponentScan;
+
+@SpringBootApplication
+@ComponentScan(basePackages = "io.dapr.spring.boot.cloudconfig.config")
```
@lony2003 here Spring usually creates an annotation to enable a feature.. is there no Spring Cloud annotation to enable access to all config features?

lony2003 replied:
Yeah, actually it doesn't depends on any Spring Cloud features, it can be a dependent library cause it's just a core of SpringBoot feature. Named Cloud Config just because this feature is usually bandled with other Spring Cloud components and requires an rpc connection

salaboy commented on dapr-spring/dapr-spring-cloudconfig/src/test/resources/application.yaml:
```diff
@@ -0,0 +1,21 @@
+    import:
+      - dapr:secret:democonfig/multivalue-properties?type=doc
+      - dapr:secret:democonfig/dapr.spring.democonfigsecret.singlevalue?type=value
+      #- dapr:secret:democonfigmulti/value1?type=value
```
@lony2003 we don't need to keep the comments.

lony2003 replied:
Removed.

salaboy commented on daprdocs/content/en/java-sdk-docs/spring-boot/_index.md:
````diff
@@ -323,6 +323,146 @@ daprWorkflowClient.raiseEvent(instanceId, "MyEvenet", event);
+dapr.cloudconfig.wait-sidecar-retries=3
+```
+
+In Dapr Cloud Config component, we support two ways to imprt config: Secret Store API and Configuration API.
````
@lony2003 typo on import

lony2003 replied:
I think I have listed below. Does it needs to be clearify at this place?

lony2003 committed:
style(springboot cloudconfig): Clean comments

Signed-off-by: lony2003 <zhangke200377@outlook.com>

lony2003 committed:
docs(springboot cloudconfig): Fix typo of import

Signed-off-by: lony2003 <zhangke200377@outlook.com>

lony2003 committed:
test(springboot cloudconfig): Fix test coverage

add a test of Dapr Secret getBulkSecret method

Signed-off-by: lony2003 <zhangke200377@outlook.com>

lony2003 commented:
Has add a test of bulk secret, not sure if it's the cause of jacoco coverage failed, but I can't add more coverage as it's a little difficult to test error situation in bootstrap stage (springboot application will crash and test will be failed)

salaboy commented:
@lony2003 this is amazing! thank so much for all the work that you put on this PR! 
I really appreciate the demo application in the spring boot examples, but I have one more ask. Can you make sure that this also works on Kubernetes? 
I want to make sure that people can understand how this will work on Kubernetes when compared with the local version.. Following something like this: https://github.com/dapr/java-sdk/tree/master/spring-boot-examples/kubernetes 
what are the steps that people will need to perform on a cluster to configure Spring Cloud Config inside Kubernetes, assuming that the config parameters are in Kubernetes ConfigMaps or Secrets. Does this make sense?

lony2003 commented:
@salaboy thank you for review.

> Can you make sure that this also works on Kubernetes? 

Yes, as long as you configure the secret store components and configuration components in your dapr runtime. Dapr Cloud Config just a client that query the secret store and configuration using java sdk.

> I want to make sure that people can understand how this will work on Kubernetes when compared with the local version.

There is no difference of what does this library do in Kubernates and local version. Just make a rpc request to the dapr runtime, and waiting to get the result.

> what are the steps that people will need to perform on a cluster to configure Spring Cloud Config inside Kubernetes, assuming that the config parameters are in Kubernetes ConfigMaps or Secrets.

if you want to use cloud config on clusters, you define a dapr component that represent a secretstore or configuration component that using Kubernates builtin feature, and have the store name configured in the springboot configuration using the schema above, and store the secret or configuration in SpringBoot style (yaml, properties, etc. like the demo above), and it works (if java-sdk didn't perform differently between Kubernates and local version of course).

I'm not that familiar with Kubernates, but if dapr runtime has the same return type compared to Kubernates and local versions, it will work at the same way.

salaboy commented:
@lony2003 I totally see that, but because Dapr usually runs on Kubernetes, it will be great to add the steps needed in the `kubernetes` directory so the same example works there. Notice how the producer and consumer apps have the instructions to create the container needed and we have the manifest to run on Kubernetes.. can you copy that for the config demo? 
I think running the Kubernetes tutorial will give you enough context to just copy what you need for the config demo.

salaboy commented:
@lony2003 if you don't feel comfortable with that.. I can spend some time after KubeCon (the whole next week) to see if I can get it working on your branch

lony2003 commented:
@salaboy I'll try to enhence that part. But may be I need your help as I really don't familiar with that😂😂😂.

I am from China, so sometimes I might use some inappropriate words when communicating in English. I hope you don't mind😃😃😃.

salaboy commented:
@lony2003 your english is more than fine! thanks so much for all the work, if you have questions don't hesitate to ask. Are you on discord right?

lony2003 commented:
> Are you on discord right?

@salaboy yeah, I'll communicate with you on discord if I meet troubles. Thanks!

lony2003 committed:
docs(cloud-config-demo): enhence Kubernetes cluster usages

Signed-off-by: lony2003 <zhangke200377@outlook.com>

lony2003 committed:
docs(cloud-config-demo): enhence Kubernetes cluster usages

Signed-off-by: lony2003 <zhangke200377@outlook.com>

lony2003 committed:
docs(cloud-config-demo): enhence Kubernetes cluster usages

Signed-off-by: lony2003 <zhangke200377@outlook.com>

lony2003 commented:
@salaboy I have tried the Kubernetes usage and everything needed is documented, it works for me.

lony2003 committed:
docs(springboot cloudconfig): A fix of example docs.

Fix the example doc to make cloud config demo works both local and kubernetes, and pass the mm doc validate

Signed-off-by: lony2003 <zhangke200377@outlook.com>

lony2003 committed:
docs(springboot cloudconfig): A fix of example docs.

Fix the example doc to make cloud config demo works both local and kubernetes

Signed-off-by: lony2003 <zhangke200377@outlook.com>

lony2003 committed:
docs(springboot cloudconfig): A fix of example docs.

Signed-off-by: lony2003 <zhangke200377@outlook.com>

salaboy reviewed (dismissed):
LGTM

salaboy commented:
@dapr/maintainers-java-sdk I believe that we can merge this PR, we will incrementally improve the quality of this PR, but I want this module to be present in the SDK. Please approve, we can fix any findings in smaller PRs.

artur-ciocanu reviewed (changes requested):
@lony2003 thanks a lot for your contribution! I have reviewed the PR and I have left quite a few comments.

The main issue that I want to highlight is that we need to make sure that we can leverage `ConfigurableBootstrapContext` and `BootstrapRegistryInitializer` to ensure we create the `DaprClient` and `DaprClientProperties` beans. As far as I saw a lot of complexity is related to the fact that we try to jump through the hoops just to get the right instances so we could load the data.

While I have mentioned `DaprClient` and `DaprClientProperties` beans, all the other dependencies like parse handler should be registered with the `ConfigurableBootstrapContext` as well, so we could look them up when needed.

Using the aforementioned approach will allow us to simplify the code and be able to expand in the future without a lot of effort.

artur-ciocanu commented:
@lony2003 also could you please sync your branch with`master` branch, your branch is out of date.

lony2003 commented:
@artur-ciocanu I am working on my graduation currently, once I have time, I‘ll resolve them.

@salaboy Thank you for approving this.

salaboy commented:
@lony2003 as soon as you have the changes done please let me know

lony2003 committed:
Merge branch 'master' into development-cloudconfig

Signed-off-by: lony2003 <35906997+lony2003@users.noreply.github.com>

cicoyle commented:
@lony2003 - can you take a look at CI?
